### PR TITLE
Property hooks

### PR DESCRIPTION
**Pull request by iluuu1994**

### CODE REVIEW

File: Zend/zend_property_hooks.c [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	if (zobj->properties) {
+		int i = 0;
+		int valid_property_count = zend_array_count(zobj->properties);
+		Bucket *bucket;
+		ZEND_HASH_MAP_FOREACH_BUCKET(zobj->properties, bucket) {
+			if (i < valid_property_count) {
+				continue;
```
Comment (iluuu1994) [inline Zend/zend_property_hooks.c:59]: Heh, infinite loop :yum:

File: Zend/tests/property_hooks/final_prop.phpt [hunk 2]
```diff
@@ -1,13 +1,12 @@
 --TEST--
-errmsg: properties cannot be final
+Final property is not allowed
 --FILE--
 <?php
 
-class test {
-    final $var = 1;
+class A {
+    public final $prop { get {} set {} }
```
Comment (Danack) [inline Zend/tests/property_hooks/final_prop.phpt:7]: The RFC says:

> A property may also be declared final. 

But the test seems to be doing the opposite?

Comment (iluuu1994) [reply Zend/tests/property_hooks/final_prop.phpt:7]: Thanks. We removed final for a short while, I added this to my TODO list.

**Comment by ikkez** [comment]
Hi @iluuu1994 
I have no idea how to get in contact and reach you through that php mailing list, but I am following the Property Hooks and the old property accessor RFC for feature a while now.

The base problem these RFCs try to solve is that we cannot intercept a public property's access, while it would be so useful. I'm here with you, but I'm thinking about a different solution that might just solve that problem in an easier way than hooks. If we'd just have a property flag to leave the public property uninitialized by default, we could fallback to the magic get/set method for it's access.
It's also possible right now, I've put it in here to show you what I mean https://3v4l.org/fQ4AG#v8.2.7
What do you think about that? Probably it's less type save by now, but maybe there's a solution for this somewhere. It would just make things so much easier to use and adapt for projects I guess.
kind regards, CK

**Comment by KapitanOczywisty** [comment]
> The base problem these RFCs try to solve is that we cannot intercept a public property's access

And asymmetric visibility, and on-set validation, and virtual properties, and allow changes in logic without breaking API, and likely more.

> I'm thinking about a different solution that might just solve that problem in an easier way than hooks

This would add new keyword for relatively low gain of single use case.

**Comment by JoyceBabu** [comment]
I am very much interested in this RFC and hope it will pass. But does this require reserving the keywords `get` and `set`? That would break every application depending on PSR-11.

**Comment by iluuu1994** [comment]
`get`/`set` aren't keywords in this RFC, functions can continue using that name.

**Comment by shaedrich** [comment]
Would it make sense, to add a `jsonSerialize` hook to the [Future scope](https://wiki.php.net/rfc/property-hooks#future_scope) for `get` hooks that don't return a scalar value?

**Comment by iluuu1994** [comment]
@javaDeveloperKid This branch is a bit outdated but functional. You can just check it out and build it like normal.

**Comment by SerafimArts** [comment]
> closed this 3 days ago

I apologize, perhaps I did not see the necessary discussion in internals/externals, but do I understand correctly that work on adding properties to the php has been cancelled? =(

Frankly, adding property support was one of the most desirable possible future innovations in the PHP.next.

**Pull request closed without merging**